airflow/models/dag_version.py

```diff
@@ -0,0 +1,162 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from __future__ import annotations
+
+import logging
+import uuid
+from typing import TYPE_CHECKING
+
+from sqlalchemy import Column, ForeignKey, Integer, func, select
+from sqlalchemy.orm import relationship
+from sqlalchemy_utils import UUIDType
+
+from airflow.models.base import Base, StringID
+from airflow.utils import timezone
+from airflow.utils.session import NEW_SESSION, provide_session
+from airflow.utils.sqlalchemy import UtcDateTime, with_row_locks
+
+if TYPE_CHECKING:
+    from sqlalchemy.orm import Session
+
+    from airflow.models.dagcode import DagCode
+    from airflow.models.serialized_dag import SerializedDagModel
+
+log = logging.getLogger(__name__)
+
+
+class DagVersion(Base):
+    """Model to track the versions of DAGs in the database."""
+
+    __tablename__ = "dag_version"
+    id = Column(UUIDType, primary_key=True, default=uuid.uuid4)
+    version_number = Column(Integer)
+    version_name = Column(StringID(), default=uuid.uuid4)
+    dag_id = Column(StringID(), ForeignKey("dag.dag_id", ondelete="CASCADE"))
+    dag_model = relationship("DagModel", back_populates="dag_versions")
+    dag_code = relationship("DagCode", back_populates="dag_version", uselist=False)
+    serialized_dag = relationship("SerializedDagModel", back_populates="dag_version", uselist=False)
+    dag_runs = relationship("DagRun", back_populates="dag_version")
+    task_instances = relationship("TaskInstance", back_populates="dag_version")
+    created_at = Column(UtcDateTime, default=timezone.utcnow)
```

Comment:
It seems to me at least some of the fields on this model should be `nullable=False`.

Reply:
Yeah. Fixed this and added unique where it should be unique. Thanks